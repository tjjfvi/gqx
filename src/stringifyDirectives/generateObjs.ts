
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
