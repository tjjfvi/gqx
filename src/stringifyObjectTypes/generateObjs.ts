
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
