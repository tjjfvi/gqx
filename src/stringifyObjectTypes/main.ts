
import { unwrapType, Context, stringifyFragResult, stringifyIds, stringifyFragEnums } from "."

interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  shallow: boolean;
  wrap: (x: string) => string;
}

interface Obj {
  type: string;
  shallowProps: Prop[];
  deepProps: Prop[];
}

export { Id, Prop, Obj };

export default (ctx: Context) => {
  const ids: Id[] = [];

  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .filter(k => !ctx.baseTypes.includes(k))
      .sort()
      .map(name => {
        const fields = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          const shallow = typeName in ctx.objectTypes;
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
        };
        return obj;
      })

  return [
    stringifyIds(ids),
    stringifyFragEnums(objs, ctx),
    ...objs.map(o => stringifyFragResult(o, ctx))
  ].join("\n\n");
}

