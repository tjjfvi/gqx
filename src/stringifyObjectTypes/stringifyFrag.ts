/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { Obj } from ".";
import { stringifyId } from "./stringifyId";
import { indent } from "../indent";

export const stringifyFrag = (obj: Obj, objs: Obj[]) => `
interface _$${obj.type}$ extends $Hkt {
  return: ${obj.type}.$<this["input"]>
}

interface _$${obj.type}<T extends $Hkt> {
  $<F extends ${obj.type}.$>(frag: F): $CallHkt<T, ${obj.type}<F>>
  $<F extends ${obj.type}.$[]>(...frag: F): $CallHkt<T, ${obj.type}<F>>
${
  obj.props.map(prop =>
    indent`${prop.id.prop}: ${
      prop.shallow ?
        `$CallHkt<T, $Wrap<$ShallowPropTypes[${stringifyId(prop.id)}]>[${stringifyId(prop.id)}]>` :
        `_$${prop.type}<$Compose<T, $WrapHkt<${stringifyId(prop.id)}>>>`
    }`
  ).join("\n")
}
}

const _$${obj.type} = <T extends $Hkt>(f: (x: $Fragment) => any): _$${obj.type}<T> => ({
  $: ((...a: $Fragment[]) => f($Fragment.merge(...a))) as any,
${
  obj.props.map(prop =>
    indent`get ${prop.id.prop}(){ return ${
      prop.shallow ?
        `f($Fragment._create(${stringifyId(prop.id)}, null))` :
        `_$${prop.type}<$Compose<T, $WrapHkt<${stringifyId(prop.id)}>>>(x => $Fragment._create(${stringifyId(prop.id)}, f(x)))`
    } },`
  ).join("\n")
}
})

export const ${obj.type} = _$${obj.type}<_$${obj.type}$>(x => x)

// @ts-ignore${/* eslint-disable-next-line max-len */""}
export type ${obj.type}<F extends ${obj.type}.$ | ${obj.type}.$[]> = F extends ${obj.type}.$<infer X> ? X : $ExpandDeep<{ [K in keyof F]: (x: ${obj.type}<F[K]>) => void }[number] extends (x: infer X) => void ? X : never>

export declare namespace ${obj.type} {
  export interface $<out T = unknown> extends $Fragment {
    [$fragTypes]: {
      _${obj.type}: T
${
  obj.supertypes.map(type =>
    indent(indent(indent`_${type}: T${
      objs
        .find(x => x.type === type)!
        .subtypes
        .filter(x => x !== obj.type)
        .map(x => ` | { __typename: ${JSON.stringify(x)} }`)
        .join("")
    }`))
  )
}
    }
  }
}
`.trim()
