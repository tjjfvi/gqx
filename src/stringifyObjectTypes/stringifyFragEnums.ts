

import { Obj, Context } from ".";
import stringifyId from "./stringifyId";

export default (objs: Obj[], ctx: Context) => {
  ctx.exports.push("$_");
  return [
    ...objs.map(obj => {
      ctx.exports.push(obj.type + "$");
      return [
        `type ${obj.type}$ =\n`,
        ...obj.shallowProps.map(p => `  | typeof ${stringifyId(p.id)}\n`),
        ...obj.deepProps.map(p => `  | $<typeof ${stringifyId(p.id)}, ${p.type}$>\n`),
      ].join("");
    }),
    `\ntype $_ =${objs.map(o => `\n  | ${o.type}$`).join("")}`,
  ].join("\n")
}
