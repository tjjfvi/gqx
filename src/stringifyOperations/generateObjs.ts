
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
