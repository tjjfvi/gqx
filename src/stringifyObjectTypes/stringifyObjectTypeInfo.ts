
import { Obj } from ".";
import stringifyId from "./stringifyId";
import { stringifyObject } from "../stringifyObject";
import indent from "../indent";
import { stringifyEnum } from "../stringifyEnum";
import stringifyInputType from "../stringifyInputType";
import { Context } from "../stringifyOperations";
import { stringifyValue } from "../stringifyValue";

const str = (str: string) => JSON.stringify(str);

export const stringifyObjectTypeInfo = (ctx: Context, objs: Obj[]) =>
  `export type $$ObjectType = ${stringifyEnum(objs.map(v => JSON.stringify(v.type)))}\n\n` +
  "const $$objectTypeInfoMap = " + stringifyObject(objs.map(obj =>
    [obj.type, [
      "new $$ObjectTypeInfo(",
      ...[
        ...[obj.shallowProps, obj.deepProps].flatMap(props =>
          !props.length ?
            "{}, {}, {}, {}" :
            [
              stringifyObject(props.map(prop => [stringifyId(prop.id), str(prop.id.prop)])),
              stringifyObject(props.map(prop => [str(prop.id.prop), stringifyId(prop.id)])),
              "null as any as " + stringifyObject(props.map(prop =>
                [stringifyId(prop.id), str(prop.type)]
              )),
              "null as any as " + stringifyObject(props.map(prop =>
                [stringifyId(prop.id), prop.wrapHKT]
              )),
            ]
        ),
        "null as any as " + stringifyObject(obj.props.map(prop =>
          [stringifyId(prop.id), stringifyInputType(ctx, prop.args)]
        )),
        stringifyObject(obj.props.map(prop =>
          [stringifyId(prop.id), stringifyObject(prop.args.map(arg =>
            [
              arg.name.value,
              JSON.stringify(arg.type.loc?.source.body.slice(arg.type.loc.start, arg.type.loc.end) ?? "")
            ]
          ))]
        )),
        stringifyObject(obj.props.map(prop =>
          [stringifyId(prop.id), stringifyObject(prop.directives.map(directive =>
            [
              directive.name.value,
              stringifyObject(directive.arguments?.map(arg =>
                [arg.name.value, stringifyValue(arg.value)]
              ) ?? [])
            ]
          ))]
        )),
        stringifyObject(obj.props.map(prop =>
          [stringifyId(prop.id), JSON.stringify(prop.loc?.source.name ?? null)]
        )) + " as const",
      ].map(str => indent`${str},`),
      ")",
    ].join("\n")]
  ))
