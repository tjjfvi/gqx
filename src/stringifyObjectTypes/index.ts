
import { Context } from "..";
import stringifyFragEnums from "./stringifyFragEnums";
import stringifyFragResult from "./stringifyFragResult";
import { stringifyObjectTypeInfo } from "./stringifyObjectTypeInfo";
import stringifyFragObject from "./stringifyFragObject";
import generateObjs from "./generateObjs";

interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  shallow: boolean;
  wrap: (x: string) => string;
  wrapHKT: string;
}

interface Obj {
  type: string;
  shallowProps: Prop[];
  deepProps: Prop[];
}

export { Id, Prop, Obj, Context };

export default (ctx: Context) => {
  const [objs] = generateObjs(ctx);

  return [
    stringifyObjectTypeInfo(objs),
    stringifyFragEnums(objs),
    ...objs.map(o => stringifyFragObject(o)),
    ...objs.map(o => stringifyFragResult(o))
  ].join("\n\n");
}

