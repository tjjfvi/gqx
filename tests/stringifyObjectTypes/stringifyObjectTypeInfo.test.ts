import { generateObjs } from "../../src/stringifyObjectTypes/generateObjs";
import { stringifyObjectTypeInfo } from "../../src/stringifyObjectTypes/stringifyObjectTypeInfo";

test("", () => {
  const [objs,, ctx] = generateObjs(utils.getOperations(["stringifyObjectTypes"]));
  const code = stringifyObjectTypeInfo(ctx, objs);
  expect(code).toParse().toMatchSnapshot();
})
