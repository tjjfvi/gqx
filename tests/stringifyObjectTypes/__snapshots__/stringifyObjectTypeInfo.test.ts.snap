// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"export type $$ObjectType = (
  | \\"BinaryTree\\"
  | \\"DeepOnly\\"
  | \\"Query\\"
  | \\"Recursive\\"
  | \\"ShallowOnly\\"
  | \\"WithId\\"
)

export const $$objectTypeInfoMap = {
  BinaryTree: new $$ObjectTypeInfo(
    \\"BinaryTree\\",
    {
      BinaryTree$id: \\"id\\",
      BinaryTree$node: \\"node\\",
    },
    {
      id: \\"BinaryTree$id\\",
      node: \\"BinaryTree$node\\",
    },
    null as any as {
      BinaryTree$id: \\"ID\\",
      BinaryTree$node: \\"String\\",
    },
    null as any as {
      BinaryTree$id: $$Identity,
      BinaryTree$node: $$Identity,
    },
    {
      BinaryTree$left: \\"left\\",
      BinaryTree$right: \\"right\\",
    },
    {
      left: \\"BinaryTree$left\\",
      right: \\"BinaryTree$right\\",
    },
    null as any as {
      BinaryTree$left: \\"BinaryTree\\",
      BinaryTree$right: \\"BinaryTree\\",
    },
    null as any as {
      BinaryTree$left: $$Optional<$$Identity>,
      BinaryTree$right: $$Optional<$$Identity>,
    },
    null as any as {
      BinaryTree$id: {},
      BinaryTree$left: {},
      BinaryTree$node: {},
      BinaryTree$right: {},
    },
    {
      BinaryTree$id: {},
      BinaryTree$left: {},
      BinaryTree$node: {},
      BinaryTree$right: {},
    },
    {
      BinaryTree$id: {},
      BinaryTree$left: {},
      BinaryTree$node: {},
      BinaryTree$right: {},
    },
    {
      BinaryTree$id: \\"GraphQL request\\",
      BinaryTree$left: \\"GraphQL request\\",
      BinaryTree$node: \\"GraphQL request\\",
      BinaryTree$right: \\"GraphQL request\\",
    } as const,
    null as any as never,
    null as any as (
      | \\"WithId\\"
    ),
  ),
  DeepOnly: new $$ObjectTypeInfo(
    \\"DeepOnly\\",
    {}, {}, {}, {},
    {
      DeepOnly$a: \\"a\\",
      DeepOnly$b: \\"b\\",
    },
    {
      a: \\"DeepOnly$a\\",
      b: \\"DeepOnly$b\\",
    },
    null as any as {
      DeepOnly$a: \\"ShallowOnly\\",
      DeepOnly$b: \\"ShallowOnly\\",
    },
    null as any as {
      DeepOnly$a: $$Identity,
      DeepOnly$b: $$Array<$$Optional<$$Identity>>,
    },
    null as any as {
      DeepOnly$a: {},
      DeepOnly$b: {},
    },
    {
      DeepOnly$a: {},
      DeepOnly$b: {},
    },
    {
      DeepOnly$a: {},
      DeepOnly$b: {},
    },
    {
      DeepOnly$a: \\"GraphQL request\\",
      DeepOnly$b: \\"GraphQL request\\",
    } as const,
    null as any as never,
    null as any as never,
  ),
  Query: new $$ObjectTypeInfo(
    \\"Query\\",
    {
      Query$oneDirective: \\"oneDirective\\",
    },
    {
      oneDirective: \\"Query$oneDirective\\",
    },
    null as any as {
      Query$oneDirective: \\"String\\",
    },
    null as any as {
      Query$oneDirective: $$Optional<$$Identity>,
    },
    {
      Query$multDirectivesArgs: \\"multDirectivesArgs\\",
      Query$multDirectivesNoArgs: \\"multDirectivesNoArgs\\",
      Query$multDirectivesNullArgs: \\"multDirectivesNullArgs\\",
      Query$noDirectives: \\"noDirectives\\",
    },
    {
      multDirectivesArgs: \\"Query$multDirectivesArgs\\",
      multDirectivesNoArgs: \\"Query$multDirectivesNoArgs\\",
      multDirectivesNullArgs: \\"Query$multDirectivesNullArgs\\",
      noDirectives: \\"Query$noDirectives\\",
    },
    null as any as {
      Query$multDirectivesArgs: \\"BinaryTree\\",
      Query$multDirectivesNoArgs: \\"ShallowOnly\\",
      Query$multDirectivesNullArgs: \\"DeepOnly\\",
      Query$noDirectives: \\"Recursive\\",
    },
    null as any as {
      Query$multDirectivesArgs: $$Identity,
      Query$multDirectivesNoArgs: $$Optional<$$Identity>,
      Query$multDirectivesNullArgs: $$Identity,
      Query$noDirectives: $$Array<$$Identity>,
    },
    null as any as {
      Query$multDirectivesArgs: {},
      Query$multDirectivesNoArgs: {},
      Query$multDirectivesNullArgs: {},
      Query$noDirectives: {},
      Query$oneDirective: {},
    },
    {
      Query$multDirectivesArgs: {},
      Query$multDirectivesNoArgs: {},
      Query$multDirectivesNullArgs: {},
      Query$noDirectives: {},
      Query$oneDirective: {},
    },
    {
      Query$multDirectivesArgs: {
        scalarArg: {
          a: \\"test\\",
          b: 12345,
        },
        inputArg: {
          a: {
            x: 5,
            y: [
              1,
              2,
              3,
            ],
            z: {
              x: 5,
            },
          },
        },
        scalarArrArg: {
          a: [
            \\"a\\",
            \\"b\\",
            \\"c\\",
            null,
          ],
        },
        inputArrArg: {
          a: [
            {
              x: 5,
            },
            {
              x: 6,
            },
            {
              y: [
                1,
                2,
                3,
              ],
            },
          ],
        },
      },
      Query$multDirectivesNoArgs: {
        noArgs: {},
        scalarArg: {},
        inputArg: {},
        scalarArrArg: {},
        inputArrArg: {},
      },
      Query$multDirectivesNullArgs: {
        scalarArg: {
          a: null,
          b: null,
        },
        inputArg: {
          a: null,
        },
        scalarArrArg: {
          a: null,
        },
        inputArrArg: {
          a: null,
        },
      },
      Query$noDirectives: {},
      Query$oneDirective: {
        scalarArg: {
          x: \\"string\\",
        },
      },
    },
    {
      Query$multDirectivesArgs: \\"GraphQL request\\",
      Query$multDirectivesNoArgs: \\"GraphQL request\\",
      Query$multDirectivesNullArgs: \\"GraphQL request\\",
      Query$noDirectives: \\"GraphQL request\\",
      Query$oneDirective: \\"GraphQL request\\",
    } as const,
    null as any as never,
    null as any as never,
  ),
  Recursive: new $$ObjectTypeInfo(
    \\"Recursive\\",
    {
      Recursive$id: \\"id\\",
    },
    {
      id: \\"Recursive$id\\",
    },
    null as any as {
      Recursive$id: \\"ID\\",
    },
    null as any as {
      Recursive$id: $$Identity,
    },
    {
      Recursive$next: \\"next\\",
    },
    {
      next: \\"Recursive$next\\",
    },
    null as any as {
      Recursive$next: \\"Recursive\\",
    },
    null as any as {
      Recursive$next: $$Identity,
    },
    null as any as {
      Recursive$id: {},
      Recursive$next: {},
    },
    {
      Recursive$id: {},
      Recursive$next: {},
    },
    {
      Recursive$id: {},
      Recursive$next: {},
    },
    {
      Recursive$id: \\"GraphQL request\\",
      Recursive$next: \\"GraphQL request\\",
    } as const,
    null as any as never,
    null as any as (
      | \\"WithId\\"
    ),
  ),
  ShallowOnly: new $$ObjectTypeInfo(
    \\"ShallowOnly\\",
    {
      ShallowOnly$a: \\"a\\",
      ShallowOnly$b: \\"b\\",
    },
    {
      a: \\"ShallowOnly$a\\",
      b: \\"ShallowOnly$b\\",
    },
    null as any as {
      ShallowOnly$a: \\"String\\",
      ShallowOnly$b: \\"Number\\",
    },
    null as any as {
      ShallowOnly$a: $$Identity,
      ShallowOnly$b: $$Identity,
    },
    {}, {}, {}, {},
    null as any as {
      ShallowOnly$a: {},
      ShallowOnly$b: {},
    },
    {
      ShallowOnly$a: {},
      ShallowOnly$b: {},
    },
    {
      ShallowOnly$a: {},
      ShallowOnly$b: {},
    },
    {
      ShallowOnly$a: \\"GraphQL request\\",
      ShallowOnly$b: \\"GraphQL request\\",
    } as const,
    null as any as never,
    null as any as never,
  ),
  WithId: new $$ObjectTypeInfo(
    \\"WithId\\",
    {
      WithId$id: \\"id\\",
    },
    {
      id: \\"WithId$id\\",
    },
    null as any as {
      WithId$id: \\"ID\\",
    },
    null as any as {
      WithId$id: $$Identity,
    },
    {}, {}, {}, {},
    null as any as {
      WithId$id: {},
    },
    {
      WithId$id: {},
    },
    {
      WithId$id: {},
    },
    {
      WithId$id: \\"GraphQL request\\",
    } as const,
    null as any as (
      | \\"BinaryTree\\"
      | \\"Recursive\\"
    ),
    null as any as never,
  ),
}"
`;
